checks/ec2.go

```diff
@@ -0,0 +1,415 @@
+		},
+		[]string{"region"},
+	)
+)
+
+func init() {
+	prometheus.MustRegister(prometheusCount, prometheusFailCount, prometheusPassCount, prometheusResponseTime, prometheusStartupTime, prometheusTerminateTime)
+}
+
+type EC2Checker struct {
+	kommons *kommons.Client `yaml:"-" json:"-"`
+}
+
+func (c *EC2Checker) SetClient(client *kommons.Client) {
+	c.kommons = client
+}
+
+func (c EC2Checker) GetClient() *kommons.Client {
+	return c.kommons
+}
+
+// Run: Check every entry from config according to Checker interface
+// Returns check result and metrics
+func (c *EC2Checker) Run(canary v1.Canary) []*pkg.CheckResult {
+	var results []*pkg.CheckResult
+	if canary.Spec.EC2.AMI != "" {
+		results = append(results, c.Check(canary, canary.Spec.EC2))
+	}
+	return results
+}
+
+// Type: returns checker type
+func (c *EC2Checker) Type() string {
+	return "ec2"
+}
+
+func (c *EC2Checker) Check(canary v1.Canary, extConfig external.Check) *pkg.CheckResult {
+	check := extConfig.(v1.EC2Check)
+	prometheusCount.WithLabelValues(check.Region).Inc()
+	namespace := canary.Namespace
+
+	kommons := c.GetClient()
+	if kommons == nil {
+		return HandleFail(check, "Kommons client not configured for ec2 checker")
+	}
+	_, ak, err := kommons.GetEnvValue(check.AccessKeyID, namespace)
+	if err != nil {
+		return HandleFail(check, fmt.Sprintf("Could not parse EC2 access key: %v", err))
+	}
+	_, sk, err := kommons.GetEnvValue(check.SecretKey, namespace)
+	if err != nil {
+		return HandleFail(check, fmt.Sprintf("Could not parse EC2 secret key: %v", err))
+	}
+	tr := &http.Transport{
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: false},
+	}
+	if check.SkipTLSVerify {
+		tr = &http.Transport{
+			TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+		}
+	}
+	cfg, err := config.LoadDefaultConfig(context.TODO(),
+		config.WithCredentialsProvider(credentials.NewStaticCredentialsProvider(ak, sk, "")),
+		config.WithRegion(check.Region),
+		config.WithHTTPClient(&http.Client{Transport: tr}),
+	)
+
+	var ami *string
+	if check.AMI == "" {
+		ssmClient := ssm.NewFromConfig(cfg)
+		arnLookupInput := &ssm.GetParameterInput{Name: aws.String("/aws/service/ami-amazon-linux-latest/amzn2-ami-hvm-x86_64-gp2")}
+		arnLookupOutput, err := ssmClient.GetParameter(context.TODO(), arnLookupInput)
+		if err != nil {
+			return HandleFail(check, fmt.Sprintf("Could not look up amazon image arn: %v", err))
+		}
+		ami = arnLookupOutput.Parameter.Value
+	} else {
+		ami = aws.String(check.AMI)
+	}
+
+	client := ec2.NewFromConfig(cfg)
+	idString := fmt.Sprintf("%v/%v/%v", canary.ClusterName, namespace, canary.Name)
+
+	describeInput := &ec2.DescribeInstancesInput{
+		Filters: []types.Filter{
+			{
+				Name: aws.String("tag:type"),
+				Values: []string{
+					"canary-checker",
+				},
+			},
+			{
+				Name: aws.String("tag:owner"),
+				Values: []string{
+					idString,
+				},
+			},
+			{
+				Name: aws.String("instance-state-name"),
+				Values: []string{
+					"running",
+				},
```

Comment:
I might be missing something but In case an instance is pending or any other state than running for a long time shouldn't we fail the test instead of terminating it. Maybe we wait for it to be running for a specific time period if it is not running until then we fail and start to create a new instance?

Reply:
this is a pre-run cleanup that terminates instances in the case that a previous run created an instance and then failed to terminate them.  Previously created pending instances should probably also be cleaned up.

Reply:
yeah they need to be cleaned for sure... but I think we should be failing in case the instance is running for a long time or if it is stuck in any other state I would terminate it and fail the test
@moshloop wdyt?

Reply:
That is what I am doing?  If the instances do not start or stop within five minutes then the test fails and exits.  The initial snippet is to handle the case where that has happened in the past and now there instances from a previous test that are still running.  The test will fail again should terminating these instances fail before a new instance is started.

Reply:
Yeah... but that is hard-coded 5 minutes... maybe that can be more flexible depending on the particular ec2Check?
Like in Junit we wait until 2*interval after that we terminate and delete. Also, can we extract them as constants as well?

Reply:
I have added a timeout input